// foundation/dist/src/dates/datetime/datetime.js
import moment from "moment";
import 'moment-timezone';
import { Timezone } from "./../../geography/geography.module";
export class DateTime {
    constructor(year, month, date, hours = 0, minutes = 0, seconds = 0, ms = 0, timezone = Timezone.UTC()) {
        const utcDate = new Date(Date.UTC(year, month, date, hours, minutes, seconds, ms));
        this.date = moment.utc().toDate();
        this.tz = timezone;
    }
    static FromDate(date, timezone) {
        return new DateTime(date.getFullYear(), date.getMonth(), date.getDate(), date.getHours(), date.getMinutes(), date.getSeconds(), date.getMilliseconds(), timezone);
    }
    static Now(timezone = Timezone.UTC()) {
        return DateTime.FromDate(new Date(), timezone);
    }
    static FromTimestamp(timestamp, timezone = Timezone.UTC()) {
        return DateTime.FromDate(timestamp.value(), timezone);
    }
    addDays(numDays) {
        return DateTime.FromDate(this.getMomentDate().add(numDays, 'days').toDate(), this.timezone());
    }
    addHours(numHours) {
        return DateTime.FromDate(this.getMomentDate().add(numHours, 'hours').toDate(), this.timezone());
    }
    addMinutes(numMinutes) {
        return DateTime.FromDate(this.getMomentDate().add(numMinutes, 'minutes').toDate(), this.timezone());
    }
    addMonths(numMonths) {
        return DateTime.FromDate(this.getMomentDate().add(numMonths, 'M').toDate(), this.timezone());
    }
    addYears(numYears) {
        return DateTime.FromDate(this.getMomentDate().add(numYears, 'y').toDate(), this.timezone());
    }
    day() {
        const utc = this.getMomentDate();
        const tzConverted = utc.tz(this.timezone().id());
        return Number(tzConverted.format("D"));
    }
    isAfter(suspect) {
        let isAfter = false;
        if (suspect instanceof DateTime) {
            const other = suspect;
            isAfter = this.getMomentDate().isAfter(other.getMomentDate());
        }
        return isAfter;
    }
    isBefore(suspect) {
        let isBefore = false;
        if (suspect instanceof DateTime) {
            const otherCreated = suspect;
            isBefore = this.getMomentDate().isBefore(otherCreated.getMomentDate());
        }
        return isBefore;
    }
    equals(suspect) {
        let isEqual = false;
        if (suspect instanceof DateTime) {
            const other = suspect;
            isEqual = this.getMomentDate().isSame(other.getMomentDate()) && this.timezone().equals(other.timezone());
        }
        return isEqual;
    }
    month() {
        const utc = this.getMomentDate();
        const tzConverted = utc.tz(this.timezone().id());
        return Number(tzConverted.format("M"));
    }
    subtractDays(numDays) {
        return DateTime.FromDate(this.getMomentDate().subtract(numDays, 'd').toDate(), this.timezone());
    }
    subtractHours(numHours) {
        return DateTime.FromDate(this.getMomentDate().subtract(numHours, 'h').toDate(), this.timezone());
    }
    subtractMinutes(numMinutes) {
        return DateTime.FromDate(this.getMomentDate().subtract(numMinutes, 'm').toDate(), this.timezone());
    }
    subtractMonths(numMonths) {
        return DateTime.FromDate(this.getMomentDate().subtract(numMonths, 'M').toDate(), this.timezone());
    }
    subtractYears(numYears) {
        return DateTime.FromDate(this.getMomentDate().subtract(numYears, 'y').toDate(), this.timezone());
    }
    timezone() {
        return this.tz;
    }
    toUtc() {
        return DateTime.FromDate(this.value(), Timezone.UTC());
    }
    toTimezone(timezone) {
        return DateTime.FromDate(this.value(), timezone);
    }
    utcString() {
        return this.getMomentDate().format();
    }
    value() {
        const utc = this.getMomentDate();
        const tzConverted = utc.tz(this.timezone().id());
        return new Date(this.utcString());
    }
    year() {
        const utc = this.getMomentDate();
        const tzConverted = utc.tz(this.timezone().id());
        return Number(this.getMomentDate().format("YYYY"));
    }
    getMomentDate() {
        return moment.utc(this.date);
    }
}
