// foundation/src/dates/timestamp/timestamp.ts

import moment, { Moment } from "moment";
import 'moment-timezone';
import { DateException } from './../exceptions/date.exception';
import { TimestampInterface } from "./timestamp.interface";
import { Equatable } from './../../common/common.module';


/**
 * Timestamp
 * 
 * Timestamp represents a moment in time in the form of a date and time.
 */

export class Timestamp implements TimestampInterface, Equatable {
    private date: Date;

    /**
     * Creates a Timestamp object.
     * @param value Date
     * @throws DateException when the date is invalid.
     */

    constructor(year: number, month: number, date: number, hours: number = 0, minutes: number = 0, seconds: number = 0, ms: number = 0) {
        this.date = new Date(year, month, date, hours, minutes, seconds, ms);
    }

    /**
     * FromDate()
     * 
     * creates a Timestamp instance from a Date object.
     * @param date The date to create a timestamp from
     */
    public static FromDate(date: Date): Timestamp {
        return new Timestamp(date.getFullYear(), date.getMonth(), date.getDate(), date.getHours(), date.getMinutes(), date.getSeconds(), date.getMilliseconds());
    }

    /**
     * Now()
     * 
     * Creates a Timestamp for the current moment in UTC format.
     */

    public static Now(): Timestamp {
        return Timestamp.FromDate(new Date(Date.UTC(0, 0, 0, 0, 0, 0, 0)));
    }

    /**
     * addDays()
     * 
     * addDays() adds numDays days to the current timestamp.
     * @param numDays The number of days to add.
     */

    public addDays(numDays: number): Timestamp {
        return Timestamp.FromDate(this.toMomentObject().add(numDays, 'days').toDate());
    }

    /**
     * addHours()
     * 
     * addHours() adds numHours hours to the timestamp.
     * @param numHours The number of hours to add.
     */

    public addHours(numHours: number): Timestamp {
        return Timestamp.FromDate(this.toMomentObject().add(numHours, 'hours').toDate());
    }

    /**
     * addMinutes()
     * 
     * addMinutes() adds numMinutes minutes to the current timestamp.
     * @param numMinutes The number of minutes to add to the timestamp.
     */

    public addMinutes(numMinutes: number): Timestamp {
        return Timestamp.FromDate(this.toMomentObject().add(numMinutes, 'minutes').toDate());
    }

    /**
     * addMonths()
     * 
     * addMonths() adds numMonths months to the current timestamp.
     * @param numMonths The number of months to add.
     */

    public addMonths(numMonths: number): Timestamp {
        return Timestamp.FromDate(this.toMomentObject().add(numMonths, 'months').toDate());
    }

    /**
     * addYears()
     * 
     * addYears() adds numYears years to the current timestamp.
     * @param numYears The number of years to add.
     */

    public addYears(numYears: number): Timestamp {
        return Timestamp.FromDate(this.toMomentObject().add(numYears, 'years').toDate());
    }

    /**
     * day()
     * 
     * day() gets the day of the month of the DateOfBirth.
     * @returns a number between 1 and 31
     */

    public day(): number {
        return Number(this.toMomentObject().format("D"));
    }

    /**
     * isAfter()
     * 
     * isAfter() compares the Created instance to the suspect, to determine if the suspect is after the instance.
     * @param suspect The suspect to be compared.
     */

    public isAfter(suspect: any): boolean {
        let isAfter = false;

        if (suspect instanceof Timestamp) {
            const other = suspect as Timestamp;
            isAfter = this.toMomentObject().isAfter(other.toMomentObject());
        }

        return isAfter;
    }

    /**
     * isBefore()
     * 
     * isBefore() compares the instance to the suspect, to determine if the instance is before the suspect.
     * @param suspect The suspect to be compared to.
     */

    public isBefore(suspect: any): boolean {
        let isBefore = false;

        if (suspect instanceof Timestamp) {
            const otherCreated = suspect as Timestamp;
            isBefore = this.toMomentObject().isBefore(otherCreated.toMomentObject());
        }

        return isBefore;
    }

    /**
     * equals()
     * 
     * equals() compares the Created instance to a suspect, to determine if they are equal.
     * @param suspect The Created object to be compared.
     */

    public equals(suspect: any): boolean {
        let isEqual = false;

        if (suspect instanceof Timestamp) {
            const other = suspect as Timestamp;
            isEqual = this.toMomentObject().isSame(other.toMomentObject());
        }

        return isEqual;
    }

    /**
     * month()
     * 
     * month() gets the month part of the DateTime.
     * @returns A numeric value (Jan = 1, Dec = 12) representing the month of the year.
     */

    public month(): number {
        // Monent counts months from 0, meaning Jan = 0 ... Dec = 11. So, we add 1 to get the correct month.
        return Number(this.toMomentObject().month());
    }



    /**
     * subtractDays()
     * 
     * subtractDays() subtracts numDays days from the current timestamp.
     * @param numDays The number of days to subtract.
     */

    public subtractDays(numDays: number): Timestamp {
        return Timestamp.FromDate(this.toMomentObject().subtract(numDays, 'days').toDate());
    }

    /**
     * subtractHours()
     * 
     * subtractHours() subtracts numHours hours from the timestamp.
     * @param numHours The number of hours to add.
     */

    public subtractHours(numHours: number): Timestamp {
        return Timestamp.FromDate(this.toMomentObject().subtract(numHours, 'h').toDate());
    }

    /**
     * subtractMinutes()
     * 
     * subtractMinutes() subtracts numMinutes minutes from the current timestamp.
     * @param numMinutes The number of minutes to subtract from the timestamp.
     */

    public subtractMinutes(numMinutes: number): Timestamp {
        return Timestamp.FromDate(this.toMomentObject().subtract(numMinutes, 'm').toDate());
    }

    /**
     * subtractMonths()
     * 
     * subtractMonths() subtracts numMonths months from the current timestamp.
     * @param numMonths The number of months to subtract.
     */

    public subtractMonths(numMonths: number): Timestamp {
        return Timestamp.FromDate(this.toMomentObject().subtract(numMonths, 'months').toDate());
    }

    /**
     * subtractYears()
     * 
     * subtractYears() subtracts numYears years from the current timestamp.
     * @param numYears The number of years to subtract.
     */

    public subtractYears(numYears: number): Timestamp {
        return Timestamp.FromDate(this.toMomentObject().subtract(numYears, 'y').toDate());
    }

    /**
     * utcString()
     * 
     * utcString() gets a UTC string for a timestamp.
     */

    public utcString(): string {
        return this.toMomentObject().format();
    }

    /**
     * value()
     * 
     * value() gets the value of the timestamp.
     */

    public value(): Date {
        return new Date(this.utcString()); 
    }

    /**
     * year()
     * 
     * year() gets the year portion of the date of birth.
     * 
     * @returns number
     */

    public year(): number {
        return Number(this.toMomentObject().format("YYYY"));
    }

    // Helpers

    /**
     * converts a date to a Moment object.
     * @param date The date to convert.
     */

    private toMomentObject(): Moment {
        return moment.utc(this.date);
    }


}
